### PR TITLE
[MPS] implement resize bilinear operation

### PR DESCRIPTION
**Pull request by fujunwei**
fix https://github.com/intel/webml-polyfill/issues/339.

### CODE REVIEW

**Comment by fujunwei** [comment]
Predict time:
WebGL  [193ms](https://drive.google.com/open?id=11AUSXuy6mSqrRN6GME31qtsD8CTCnVB7)
MPS .     [118ms](https://drive.google.com/open?id=1qvQYcKLhcsCcmz00WG1oU6WGhvxaQaXl)

**Comment by fujunwei** [comment]
PTAL, thanks.

**Comment by huningxin** [comment]
Thanks for the PR @fujunwei !

Could you please please split this PR into two:
1. implement resize bilinear op
2. compile deeplab depthwise conv

It would be easy to review and maintain in the future. What do you think?

**Comment by fujunwei** [comment]
Done, thanks.

File: services/ml/compilation_impl_mac_mps.mm [hunk 1]
```diff
@@ -563,4 +563,41 @@ bool CompileFullyConnected(std::map<uint32_t, MPSNNImageNode*>& image_nodes,
...
+  const OperandMac& output_operand = operands[operation.outputs[0]];
+  if (output_operand.dimensions.size() != 4) {
+    LOG(ERROR) << "Input and output must be 4-D tensor.";
+    return false;
+  }
+
+  const OperandMac& input_operand = operands[operation.inputs[0]];
+  if (output_operand.dimensions[2] % input_operand.dimensions[2] != 0 ||
+      output_operand.dimensions[1] % input_operand.dimensions[1] != 0) {
+    LOG(ERROR) << "The upsampling factor for the x/y dimension is integer.";
```
Comment (huningxin) [inline services/ml/compilation_impl_mac_mps.mm]: It looks like MPS only support integer scale factor. So here the log message seems opposite. should be "not integer" or "must be integer".

File: services/ml/compilation_impl_mac_mps.mm [hunk 2]
```diff
@@ -563,4 +563,41 @@ bool CompileFullyConnected(std::map<uint32_t, MPSNNImageNode*>& image_nodes,
...
+  NSUInteger scale_factorX =
+      output_operand.dimensions[2] / input_operand.dimensions[2];
+  NSUInteger scale_factorY =
+      output_operand.dimensions[1] / input_operand.dimensions[1];
+
+  MPSCNNUpsamplingBilinearNode* bilinear_scale_node =
+      [[MPSCNNUpsamplingBilinearNode alloc]
+               initWithSource:image_nodes[operation.inputs[0]]
+          integerScaleFactorX:scale_factorX
+          integerScaleFactorY:scale_factorY
```
Comment (huningxin) [inline services/ml/compilation_impl_mac_mps.mm]: If MPS only support integer scale factor, we should document that.

File: services/ml/compilation_impl_mac.mm [hunk 3]
```diff
@@ -250,8 +250,11 @@
     } else if (type == mojom::FULLY_CONNECTED) {
       success = CompileFullyConnected(mps_image_nodes_, operation, operands_,
                                       values_, memory_);
+    } else if (type == mojom::RESIZE_BILINEAR) {
+      success = CompileBilinearScale(mps_image_nodes_, operation, operands_,
+                                     values_, memory_);
     } else {
-      DLOG(ERROR) << "Operation is not supported";
+      NOTREACHED();
```
Comment (huningxin) [inline services/ml/compilation_impl_mac.mm]: Does it mean GPU process will crash when the op is not supported? Should we report an error instead?

File: services/ml/compilation_impl_mac_mps.mm (hunk 2 above)
Comment (fujunwei) [reply services/ml/compilation_impl_mac_mps.mm]: it has been implemented in my local patch with MPSNNResizeBilinear that isn't NNGraph node, i will send a new pr for it to support 513 Deeplab model.

File: services/ml/compilation_impl_mac.mm (hunk 3 above)
Comment (fujunwei) [reply services/ml/compilation_impl_mac.mm]: Done with `success = false`

**Comment by huningxin** [comment]
@fujunwei , I'd suggest we split this PR into three PRs (not commits):
1. support atrous convolution (https://github.com/intel/webml-polyfill/issues/360)
2. support resize bilinear (https://github.com/intel/webml-polyfill/issues/339) with the decision of which API to use (https://github.com/intel/webml-polyfill/issues/447)
3. workaround depthwise convolution issue (https://github.com/intel/webml-polyfill/issues/446)

The 1 and 2 features are key for DeepLab model. So I believe with the two landed, we can run the DeepLab example with decent performance. Let's leave 3 for more investigation which looks like a MPS API issue. What do you think?

**Comment by fujunwei** [comment]
It also make sense, PTAL again, thanks.

**Comment by huningxin** [comment]
Looks good. Thanks!

**Pull request merged**